app/tests/src/ingestion/test_markdown_tree.py

```diff
@@ -0,0 +1,345 @@
+    assert len(list_nodes) == 4
+    assert len(list_nodes[0].children) == 3
+    assert len(list_nodes[1].children) == 4
+    # Check sublist ListItem on line 42
+    assert tree["LI_42"].render() == "* Item H1>H3.L1.subL.1\n"
+    # Indented correctly when entire list is rendered
+    assert tree["L_40"].render() == (
+        "* Item H1>H3.L1.1\n"  #
+        "* Item H1>H3.L1.2\n"  #
+        "  * Item H1>H3.L1.subL.1\n"  #
+        "  * Item H1>H3.L1.subL.2\n"  #
+        "  * Item H1>H3.L1.subL.3\n"  #
+        "* Item H1>H3.L1.1\n"  #
+        "* Item H1>H3.L1.2\n"  #
+    )
+
+    table_nodes = tree.find_all(match=lambda n: n.data_type == "Table")
+    assert len(table_nodes) == 1
+    assert len(table_nodes[0].children) == 3
+    assert "| H3.2.T1: row 1, col 1 |" in table_nodes[0].first_child().render()
+    assert "| H3.2.T1: row 3, col 4 |" in table_nodes[0].last_child().render()
+
+    paragraph_nodes = tree.find_all(match=lambda n: n.data_type == "Paragraph")
+    assert len(paragraph_nodes) == 24
+
+
+def assert_tree_structure(tree: Tree):
+    "These assertions should be true after any tree modification"
+    tree_descr = describe_tree(tree)
+
+    assert tree_descr["counts"]["Document"] == 1
+    assert tree_descr["counts"]["Paragraph"] == 24
+    assert tree_descr["counts"]["Heading"] == 5
+    assert tree_descr["counts"]["List"] == 4
+    assert tree_descr["counts"]["ListItem"] == 13
+    assert tree_descr["counts"]["Table"] == 1
+    assert tree_descr["counts"]["TableRow"] == 3
+
+    children_of = tree_descr["children"]
+    assert "Paragraph" in children_of["Document"]
+    assert children_of["List"] == {"ListItem"}
+    # ListItem can have a child Paragraph or List (i.e., nested list)
+    assert children_of["ListItem"] == {"List", "Paragraph"}
+    assert children_of["Table"] == {"TableRow"}
+    return tree_descr
+
+
+def test_create_markdown_tree(markdown_text):
+    _tree = create_markdown_tree(markdown_text)
+    tree_descr = assert_tree_structure(_tree)
+    parent_of = tree_descr["parents"]
+    # These are true initially but will change after tree preparation
+    assert parent_of["Heading"] == {"Document"}
+    assert parent_of["RawText"] == {
+        "Heading",
+        "Paragraph",
+        "Emphasis",
+        "Strong",
+        "Link",
+        "TableCell",
+    }
+
+    children_of = tree_descr["children"]
+    assert "Heading" in children_of["Document"]
+    assert children_of["Heading"] == {"RawText"}
+    assert children_of["Paragraph"] == {"RawText", "Emphasis", "Strong", "Link"}
+    assert children_of["TableRow"] == {"TableCell"}
+    assert children_of["TableCell"] == {"RawText"}
+
+    doc_node = _tree.children[0]
+    assert len(doc_node.children) == 40
+    assert_content(_tree)
+    assert len(tokens_vs_tree_mismatches(_tree)) == 0
+
+
+@pytest.fixture
+def tree(markdown_text):
+    return create_markdown_tree(markdown_text)
+
+
+def test_markdown_tokens_as_json(markdown_text):
```

Comment:
Quick note while I review the rest, you may have meant to move this to `test_markdown_tree.py`?